aws-s3-transfer-manager/src/middleware/limit/concurrency/future.rs

```diff
@@ -3,35 +3,102 @@
-        _permit: OwnedWorkPermit
+        state: State<Oneshot<S, Request>>
     }
 }
 
-impl<T> ResponseFuture<T> {
-    pub(crate) fn new(inner: T, _permit: OwnedWorkPermit) -> ResponseFuture<T> {
-        ResponseFuture { inner, _permit }
+pin_project! {
+    #[project = StateProj]
+    #[derive(Debug)]
+    enum State<F> {
+        // Polling the future from [`Service::call`]
+        Called {
+            #[pin]
+            fut: F,
+            // retain until dropped when future completes
+            _permit: OwnedWorkPermit,
+        },
+        // Polling the future from [`Scheduler::acquire_permit`]
+        AcquiringPermit {
+            #[pin]
+            permit_fut: AcquirePermitFuture
+        }
+    }
+}
+
+impl<S, Request> ResponseFuture<S, Request>
+where
+    S: Service<Request>,
+{
+    pub(crate) fn new(
+        inner: S,
+        req: Request,
+        permit_fut: AcquirePermitFuture,
+        inflight: Arc<AtomicUsize>,
+    ) -> ResponseFuture<S, Request> {
+        ResponseFuture {
+            request: Some(req),
+            svc: Some(inner),
+            inflight,
+            state: State::AcquiringPermit { permit_fut },
+        }
     }
 }
 
-impl<F, T, E> Future for ResponseFuture<F>
+impl<S, Request> Future for ResponseFuture<S, Request>
 where
-    F: Future<Output = Result<T, E>>,
+    S: Service<Request>,
+    S::Error: From<error::Error>,
 {
-    type Output = Result<T, E>;
+    type Output = Result<S::Response, S::Error>;
 
     fn poll(self: std::pin::Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Self::Output> {
-        Poll::Ready(ready!(self.project().inner.poll(cx)))
+        let mut this = self.project();
+        loop {
+            match this.state.as_mut().project() {
+                StateProj::AcquiringPermit { permit_fut } => {
+                    let res = ready!(permit_fut.poll(cx));
+                    match res {
+                        Ok(_permit) => {
+                            let req = this.request.take().expect("request set");
+                            let inflight = this.inflight.fetch_add(1, atomic::Ordering::SeqCst) + 1;
+                            tracing::trace!("in-flight requests: {inflight}");
+                            let svc = this.svc.take().expect("service set");
+                            // NOTE: because the service was (1) never polled for readiness
+                            // originally and (2) also cloned, we need to ensure it's ready now before calling it.
+                            let fut = Oneshot::new(svc, req);
+                            this.state.set(State::Called { fut, _permit });
+                        }
+                        Err(err) => return Poll::Ready(Err(err.into())),
+                    }
+                }
+                StateProj::Called { fut, .. } => {
+                    let result = ready!(fut.poll(cx));
+                    let inflight = this.inflight.fetch_sub(1, atomic::Ordering::SeqCst);
```

Comment:
Is there any chance that this future never reaches this point, and this shared atomic is never decremented? Should there be some drop logic to work around this?

Is there any chance that this future is polled again after reporting that it's ready, and this shared atomic is decremented again? Should we move to some new Completion state to avoid this?

I genuinely don't know. I've always been a bit confused over what is guaranteed to happen, or never happen, or happen exactly once in the world of Rust futures

Reply:
It's a good question, the documentation for [`poll`](https://doc.rust-lang.org/std/future/trait.Future.html#tymethod.poll) says you shouldn't poll futures once complete as they could panic, block forever, or cause other issues. 

This was rammed in last minute for my own sanity testing things, but I envision we'll want some sort of runtime/scheduler metrics that this gets tacked onto _somewhere_.